pkg/services/meta/containers.go

```diff
@@ -77,61 +113,122 @@ func (s *containerStorage) putMPTIndexes(ee []objEvent) {
+		if len(batch) > 0 {
+			err := s.db.PutChangeSet(batch, nil)
 			if err != nil {
-				// TODO define behavior with status (non-network) errors; maybe it is near #3140
-				return fmt.Errorf("HEAD object: %w", err)
+				finalErr = fmt.Errorf("put change set to DB: %w", err)
 			}
+		}
+	}()
 
-			objects[i] = h
-			return nil
-		})
-	}
+	for _, e := range ee {
+		err := isOpAllowed(s.db, e)
+		if err != nil {
+			l.Warn("skip object", zap.Stringer("oid", e.oID), zap.String("reason", err.Error()))
+			continue
+		}
 
-	err := wg.Wait()
-	if err != nil {
-		return err
-	}
+		evWithMpt := eventWithMptKVs{ev: e}
 
-	s.m.Lock()
-	defer s.m.Unlock()
-	batch := make(map[string][]byte)
+		h, err := net.Head(ctx, e.cID, e.oID)
+		if err != nil {
+			// TODO define behavior with status (non-network) errors; maybe it is near #3140
+			return fmt.Errorf("HEAD %s object: %w", e.oID, err)
+		}
 
-	for i, e := range ee {
 		commsuffix := e.oID[:]
 
 		batch[string(append([]byte{oidIndex}, commsuffix...))] = []byte{}
 		if len(e.deletedObjects) > 0 {
 			batch[string(append([]byte{deletedIndex}, commsuffix...))] = e.deletedObjects
-			err = deleteObjectsOps(batch, s.mptOpsBatch, s.db, e.deletedObjects)
+			evWithMpt.additionalKVs, err = deleteObjectsOps(batch, s.db, e.deletedObjects)
 			if err != nil {
-				return fmt.Errorf("cleaning operations for %s object: %w", e.oID, err)
+				l.Error("cleaning deleted object", zap.Stringer("oid", e.oID), zap.Error(err))
+				continue
 			}
 		}
 		if len(e.lockedObjects) > 0 {
 			batch[string(append([]byte{lockedIndex}, commsuffix...))] = e.lockedObjects
+
+			for locked := range slices.Chunk(e.lockedObjects, oid.Size) {
+				batch[string(append([]byte{lockedByIndex}, locked...))] = commsuffix
+			}
 		}
 
-		err = object.VerifyHeaderForMetadata(objects[i])
+		err = object.VerifyHeaderForMetadata(h)
 		if err != nil {
-			return fmt.Errorf("invalid %s header: %w", e.oID, err)
+			l.Error("header verification", zap.Stringer("oid", e.oID), zap.Error(err))
+			continue
 		}
 
-		fillObjectIndex(batch, objects[i])
+		res <- evWithMpt
+
+		fillObjectIndex(batch, h)
 	}
 
-	err = s.db.PutChangeSet(batch, nil)
-	if err != nil {
-		return fmt.Errorf("put change set to DB: %w", err)
+	return finalErr
+}
+
+func isOpAllowed(db storage.Store, e objEvent) error {
+	if len(e.deletedObjects) == 0 && len(e.lockedObjects) == 0 {
+		return nil
+	}
+
+	key := make([]byte, 1+oid.Size)
+
+	for obj := range slices.Chunk(e.deletedObjects, oid.Size) {
+		copy(key[1:], obj)
+
+		// delete object that does not exist
+		key[0] = oidIndex
+		_, err := db.Get(key)
+		if err != nil {
+			return fmt.Errorf("%s object-to-delete's presence check: %w", oid.ID(obj), err)
+		}
+
+		// delete object that is locked
+		key[0] = lockedByIndex
+		v, err := db.Get(key)
+		if err != nil {
+			if errors.Is(err, storage.ErrKeyNotFound) {
+				continue
+			}
+			return fmt.Errorf("%s object-to-delete's lock status check: %w", oid.ID(obj), err)
```

Comment:
as always, i wouldn't attach the input (OID) to the error context. Now it will cause a tautology in the log entry `XXX object-to-delete's lock status check: cause {"oid": "XXX", }

in the whole func

Reply:
that's the trick: `obj` is not input (well, partially it is), it is the object that is being deleted/locked. in logs `{"oid": "XXX",  ...}` part will be about LOCK/TS, not about LOCKed/TSed object